### PR TITLE
Add support for different model prediction types in DDIMInverseScheduler

### PR DESCRIPTION
**Pull request by clarencechen**
Resolve `alpha_prod_t_prev` index issue for final step of inversion
@patrickvonplaten TODO: Update numerical tests to reflect two more steps of inversion

### CODE REVIEW

**Comment by clarencechen** [comment]
@patrickvonplaten 
Here is a notebook demonstrating a working run of the Pix2PixZero pipeline with Stable Diffusion 2. I find that the value of `cross_attention_guidance_amount` needs to be lowered to 0.01 or less in order to prevent loss explosion.

https://colab.research.google.com/gist/clarencechen/4166c7d1ad29b09940a13f62437de634/ddiminversetesting.ipynb

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def __init__(
         beta_schedule: str = "linear",
         trained_betas: Optional[Union[np.ndarray, List[float]]] = None,
         clip_sample: bool = True,
-        set_alpha_to_one: bool = True,
+        set_alpha_to_zero: bool = True,
```
[the remark this answers is not in the document]
Comment (clarencechen) [reply src/diffusers/schedulers/scheduling_ddim_inverse.py:127]: Upated to deprecate argument instead.

**Comment by sayakpaul** [comment]
Thanks so much! 

What's pending in this PR?

Maybe we should mention 

> I find that the value of cross_attention_guidance_amount needs to be lowered to 0.01 or less in order to prevent loss explosion.

in the [Pix2PixZero docs](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/pix2pix_zero)?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 2]
```diff
@@ -1181,16 +1198,19 @@ def invert(
                         if lambda_auto_corr > 0:
                             for _ in range(num_auto_corr_rolls):
                                 var = torch.autograd.Variable(noise_pred.detach().clone(), requires_grad=True)
-                                l_ac = self.auto_corr_loss(var, generator=generator)
+                                var_epsilon = self.get_epsilon(var, latent_model_input.detach(), t)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1201]: Maybe add a one-liner comment to explain the rationale behind delegating this to the function? Might be helpful for future references.

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 3]
```diff
@@ -157,6 +168,7 @@ def __init__(
         self.num_inference_steps = None
         self.timesteps = torch.from_numpy(np.arange(0, num_train_timesteps).copy().astype(np.int64))
 
+    # Copy from diffusers.schedulers.scheduling_ddim.DDIMScheduler.scale_model_input
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:171]: Should be `# Copied from ...`

**Review by sayakpaul — COMMENTED** [review]
Thanks a lot for working on this! 

I see some non-trivial changes in the scheduler script. Do you mind elaborating a bit more on those?

**Comment by sayakpaul** [comment]
Sure, thanks for letting us know! Let us know if we can help in unblocking anything for you here.

**Comment by clarencechen** [comment]
@sayakpaul @patrickvonplaten In order to upload expected results for a Pix2PixZero slow test with SD 2, I need to update the following repo:
https://huggingface.co/datasets/hf-internal-testing/diffusers-images
I created a new local branch with the new data, but I need push access to the `hf-internal-testing` in order to push the branch to remote and open a PR there.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 4]
```diff
@@ -153,6 +153,8 @@ class Pix2PixInversionPipelineOutput(BaseOutput):
         >>> source_embeds = pipeline.get_embeds(source_prompts)
         >>> target_embeds = pipeline.get_embeds(target_prompts)
         >>> # the latents can then be used to edit a real image
+        >>> # when using Stable Diffusion 2 or other models that use v-prediction
+        >>> # set `cross_attention_guidance_amount` to 0.01 or less to avoid input latent gradient explosion
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:157]: Amazing this!

File: tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py [hunk 5]
```diff
@@ -366,6 +366,29 @@ def test_stable_diffusion_pix2pix_inversion(self):
 
         assert np.abs(expected_slice - image_slice.cpu().numpy()).max() < 5e-2
 
+    def test_stable_diffusion_2_pix2pix_inversion(self):
+        pipe = StableDiffusionPix2PixZeroPipeline.from_pretrained(
+            "stabilityai/stable-diffusion-2-1", safety_checker=None, torch_dtype=torch.float16
+        )
+        pipe.inverse_scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
+        pipe.inverse_scheduler = DDIMInverseScheduler.from_config(pipe.scheduler.config)
```
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:374]: Why two consecutive inverse schedulers?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py [hunk 6]
```diff
@@ -407,3 +430,45 @@ def test_stable_diffusion_pix2pix_full(self):
...
+        # numpy array of https://huggingface.co/datasets/hf-internal-testing/diffusers-images/blob/main/pix2pix/dog_2.png
+        expected_image = load_numpy(
+            "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/pix2pix/dog_2.npy"
+        )
+
+        pipe = StableDiffusionPix2PixZeroPipeline.from_pretrained(
+            "stabilityai/stable-diffusion-2-1", safety_checker=None, torch_dtype=torch.float16
+        )
+        pipe.inverse_scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
+        pipe.inverse_scheduler = DDIMInverseScheduler.from_config(pipe.scheduler.config)
```
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:444]: Same question.

**Review by sayakpaul — APPROVED** [review]
I think we're in a very good shape to merge this. Thanks so much for this important PR :)

**Comment by sayakpaul** [comment]
> @sayakpaul @patrickvonplaten In order to upload expected results for a Pix2PixZero slow test with SD 2, I need to update the following repo: https://huggingface.co/datasets/hf-internal-testing/diffusers-images I created a new local branch with the new data, but I need push access to the `hf-internal-testing` org in order to push the branch to remote and open a PR there.

Yeah feel free to open a PR to https://huggingface.co/datasets/hf-internal-testing/diffusers-images following instructions from https://huggingface.co/datasets/hf-internal-testing/diffusers-images/discussions (Click on "New pull request" and follow the instructions).

File: tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py [hunk 7]
```diff
@@ -366,6 +366,29 @@ def test_stable_diffusion_pix2pix_inversion(self):
 
         assert np.abs(expected_slice - image_slice.cpu().numpy()).max() < 5e-2
 
+    def test_stable_diffusion_2_pix2pix_inversion(self):
+        pipe = StableDiffusionPix2PixZeroPipeline.from_pretrained(
+            "stabilityai/stable-diffusion-2-1", safety_checker=None, torch_dtype=torch.float16
+        )
+        pipe.inverse_scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:373]:
```suggestion
```

We can delete this line no?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py [hunk 8]
```diff
@@ -407,3 +430,45 @@ def test_stable_diffusion_pix2pix_full(self):
 
         max_diff = np.abs(expected_image - image).mean()
         assert max_diff < 0.05
+
+    def test_stable_diffusion_2_pix2pix_full(self):
+        # numpy array of https://huggingface.co/datasets/hf-internal-testing/diffusers-images/blob/main/pix2pix/dog_2.png
+        expected_image = load_numpy(
+            "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/pix2pix/dog_2.npy"
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:437]: Indeed, could you maybe open a PR to add the numpy image? This would be very nice!

[the remark this answers is not in the document]
Comment (clarencechen) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:437]: https://huggingface.co/datasets/hf-internal-testing/diffusers-images/discussions/10

Can't currently publish due to rate limiting

File: tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py [hunk 9]
```diff
@@ -407,3 +430,45 @@ def test_stable_diffusion_pix2pix_full(self):
...
+        )
+
+        pipe = StableDiffusionPix2PixZeroPipeline.from_pretrained(
+            "stabilityai/stable-diffusion-2-1", safety_checker=None, torch_dtype=torch.float16
+        )
+        pipe.inverse_scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:443]:
```suggestion
```

File: tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py:437]: Thanks a lot for opening it - merged it!

**Comment by patrickvonplaten** [comment]
Great work @clarencechen - thanks!
